KEYCLOAK-12753 Documentation for authentication changes

mposolda opened the pull request.

mposolda committed:
KEYCLOAK-12753 Documentation for authentication changes

mposolda commented:
@andymunro Do you please have a chance to review this documentation for authentication changes? Thanks!

andymunro commented on release_notes/topics/9_0_0.adoc:
```diff
@@ -14,6 +14,16 @@ A new built-in vault provider that reads secrets from a keystore-backed Elytron
 
+== More updates to W3C WebAuthn and Authentication flows
+
+In this release, we did some usability improvements to the authentication flows. It should be more easy for end user to choose between
```
change 'more easy for end user"  to "easier for the end user"

mposolda replied:
Fixed

andymunro commented on release_notes/topics/9_0_0.adoc:
```diff
@@ -14,6 +14,16 @@ A new built-in vault provider that reads secrets from a keystore-backed Elytron
+== More updates to W3C WebAuthn and Authentication flows
+
+In this release, we did some usability improvements to the authentication flows. It should be more easy for end user to choose between
+available authentication mechanisms for two-factor authentication. It should be more intuitive to login with OTP or WebAuthn
```
change "login" to "log in"

mposolda replied:
Fixed

andymunro commented on release_notes/topics/9_0_0.adoc:
```diff
@@ -14,6 +14,16 @@ A new built-in vault provider that reads secrets from a keystore-backed Elytron
+In this release, we did some usability improvements to the authentication flows. It should be more easy for end user to choose between
+available authentication mechanisms for two-factor authentication. It should be more intuitive to login with OTP or WebAuthn
+considering the fact that user can have more OTP or WebAuthn credentials. There is also better support for passwordless WebAuthn authentication.
+Finally, we did some bugfixing related to the authentication flows.
```
change bugfixing to "work on defects"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -24,19 +24,26 @@ flow is and does.
-  to that account.  For example, if you switch the `Browser - Conditional OTP` sub-flow to `Required`, users that don't have an OTP generator configured
-  will be asked to do so.
+  For a flow to be evaluated as successful, all required elements in the flow must evaluate as successful. This means that all _Required_ elements in the flow
+  must be sequentially executed, from top to bottom, unless one the the elements causes the flow to fail. However, this is only true for the current flow.
```
duplicate "the"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -108,23 +116,8 @@ a single form for the user, like what is done for the built-in `Registration` fl
-  acts as _Disabled_. If no "Condition" execution is set, the _Conditional_ sub-flow acts as _Disabled_. If a flow contains "Condition" executions and is not set to
-  _Conditional_, the "Condition" executions are not evaluated, and can be considered functionally _Disabled_.
+Fully understanding this requires a more complete explanation of how requirements work when evaluating a flow, and this also applies to sub-flows.
+Take a look at the <<_execution-requirements, execution requirements section>> above for more details.
```
Change "Take a look at" to "Refer to"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
 
-* If the user has any WebAuthn credentials recorded, he will be able to use any of them to log in directly. This is the password-less login.
-  The user also has a drop-down allowing him to select "Password with OTP". He can do this because the "WebAuthn" execution and the "Password with OTP"
+* If the user has any WebAuthn passwordless credentials recorded, he will be able to use any of them to log in directly. This is the password-less login.
```
change "he" to "that user"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
-* If the user has any WebAuthn credentials recorded, he will be able to use any of them to log in directly. This is the password-less login.
-  The user also has a drop-down allowing him to select "Password with OTP". He can do this because the "WebAuthn" execution and the "Password with OTP"
+* If the user has any WebAuthn passwordless credentials recorded, he will be able to use any of them to log in directly. This is the password-less login.
+  The user also has a way allowing him to select "Password with OTP". He can do this because the "WebAuthn Passwordless" execution and the "Password with OTP"
```
Change " The user also has a way allowing him to select"  to "The user can instead select"
Change "He" to "The user"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
+  The user also has a way allowing him to select "Password with OTP". He can do this because the "WebAuthn Passwordless" execution and the "Password with OTP"
   flow are set to _Alternative_. Were they set to _Required_ the user would have to enter WebAuthn, password, and OTP.
-* If the user selects the "Password with OTP" from the drop-down, or if the user doesn't have any WebAuthn credentials, he will have to first enter his
+* If the user selects `Try another way` link on the screen with "WebAuthn passwordless" authentication, he can choose between "Password" and
```
Change "he" to "the user"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
   flow are set to _Alternative_. Were they set to _Required_ the user would have to enter WebAuthn, password, and OTP.
-* If the user selects the "Password with OTP" from the drop-down, or if the user doesn't have any WebAuthn credentials, he will have to first enter his
+* If the user selects `Try another way` link on the screen with "WebAuthn passwordless" authentication, he can choose between "Password" and
+  "Security Key" (WebAuthn passwordless). When selecting password, the user will need to continue and login with his OTP as well.
```
add "the" before "password"
change "login" to "log in"
change "his" to "the assigned"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
-* If the user selects the "Password with OTP" from the drop-down, or if the user doesn't have any WebAuthn credentials, he will have to first enter his
+* If the user selects `Try another way` link on the screen with "WebAuthn passwordless" authentication, he can choose between "Password" and
+  "Security Key" (WebAuthn passwordless). When selecting password, the user will need to continue and login with his OTP as well.
+  If the user doesn't have any WebAuthn credentials, he will have to first enter his
```
change "doesn't have any" to "has no"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/flows.adoc:
```diff
@@ -179,14 +172,16 @@ image:images/Passwordless-browser-login.png[]
-It is important to note that since the WebAuthn execution is set to _Alternative_ instead of _Required_, this flow will never ask the user to register a WebAuthn credential. For a user
-to have a Webauthn credential, he must have a required action added to him by an administrator. This is done first by making sure that that the `Webauthn Register`
+It is important to note that since the WebAuthn Passwordless execution is set to _Alternative_ instead of _Required_, this flow will never ask the user to register a WebAuthn credential. For a user
+to have a Webauthn credential, he must have a required action added to him by an administrator. This is done first by making sure that that the `Webauthn Register Passwordless`
```
change "he" to "that user"
delete "to him"
duplicate "that"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/webauthn.adoc:
```diff
@@ -68,11 +68,10 @@ image:images/webauthn-browser-flow-conditional-with-OTP.png[]
 ==== Authenticate with WebAuthn Authenticator
 
-After registering their WebAuthn authenticator, the user carries out the following operations :
+After registering their WebAuthn authenticator, the user carries out the following operations
```
change "their" to "a"

mposolda replied:
Fixed

andymunro commented on server_admin/topics/authentication/webauthn.adoc:
```diff
@@ -68,11 +68,10 @@ image:images/webauthn-browser-flow-conditional-with-OTP.png[]
-- Open the login form.
-- Select the credential or credentials that {project_name} must accept for login. Then click `Save` and  click `Login`.
-- The list of registered Webauthn Authenticators' labels appears. Click `Authenticate`.
+- Open the login form. User must authenticate with username and password.
```
Change "User must authenticate with"  to "The user must authenticate with a"

mposolda replied:
Fixed

andymunro commented on server_development/topics/auth-spi.adoc:
```diff
@@ -1,7 +1,7 @@
 == Authentication SPI
 
-{project_name} comes out of the box with a bunch of different authentication mechanisms: kerberos, password, and otp.
+{project_name} comes out of the box with a bunch of different authentication mechanisms: kerberos, password, otp and others.
```
change "comes out of the box with a bunch of" to "includes a range of"

mposolda replied:
Fixed

andymunro commented on server_development/topics/auth-spi.adoc:
```diff
@@ -798,7 +798,8 @@ a different credential or execution.
 If you go to the Authentication menu item and go to the Flow tab, you will be able to view the currently defined flows.
 You cannot modify built in flows, so, to add the Authenticator we've created you have to copy an existing flow or create your own.
-I'm hoping the UI is intuitive enough so that you can figure out for yourself how to create a flow and add the Authenticator.
+I'm hoping the UI is intuitive enough so that you can figure out for yourself how to create a flow and add the Authenticator. For
```
Change "I'm hoping the UI is intuitive enough so that you can figure out for yourself how to" 
to
"Our hope is that the user interface is sufficiently clear so that you can determine how to"

mposolda replied:
Fixed

andymunro reviewed (changes requested):
Some changes suggested.

mposolda committed:
KEYCLOAK-12753 Apply feedback

mposolda commented:
@andymunro Thanks a lot for the review! I've applied all the suggestions you proposed. Do you have a chance to re-review please?

I've updated the PR and used separate commit, in case you want to see just the latest changes I've updated per your feedback in separate commit. When merging later, I will squash into single commit.
